pkg/addonmanager/controllers/agentdeploy/controller.go

```diff
@@ -112,116 +141,204 @@ func (c *addonDeployController) sync(ctx context.Context, syncCtx factory.SyncCo
+		owner = metav1.NewControllerRef(addon, addonapiv1alpha1.GroupVersion.WithKind("ManagedClusterAddOn"))
+	default:
+		return nil, nil, fmt.Errorf("invalid install mode %v", installMode)
+	}
+
+	objects, err := agentAddon.Manifests(cluster, addon)
+	if err != nil {
+		meta.SetStatusCondition(&addon.Status.Conditions, metav1.Condition{
+			Type:    appliedType,
 			Status:  metav1.ConditionFalse,
 			Reason:  constants.AddonManifestAppliedReasonWorkApplyFailed,
 			Message: fmt.Sprintf("failed to get manifest from agent interface: %v", err),
 		})
-		if updateErr := utils.PatchAddonCondition(ctx, c.addonClient, managedClusterAddonCopy, managedClusterAddon); updateErr != nil {
-			return fmt.Errorf("failed to update managedclusteraddon status: %v; the err should be %v", updateErr, err)
-		}
-		return err
+		return nil, nil, err
 	}
 	if len(objects) == 0 {
-		err = deleteWork(ctx, c.workClient, clusterName, constants.DeployWorkName(addonName))
-		if err != nil {
-			return err
-		}
-		c.cache.removeCache(constants.DeployWorkName(addonName), clusterName)
-		return nil
+		return nil, nil, nil
 	}
 
-	work, _, err := newManagedManifestWorkBuilder(agentAddon.GetAgentAddonOptions().HostedModeEnabled).
-		buildManifestWorkFromObject(clusterName, managedClusterAddon, objects)
+	deployWork, hookWork, err = workBuilder.buildManifestWorkFromObject(workNamespace, addon, objects)
 	if err != nil {
-		meta.SetStatusCondition(&managedClusterAddonCopy.Status.Conditions, metav1.Condition{
-			Type:    constants.AddonManifestApplied,
+		meta.SetStatusCondition(&addon.Status.Conditions, metav1.Condition{
+			Type:    appliedType,
 			Status:  metav1.ConditionFalse,
 			Reason:  constants.AddonManifestAppliedReasonWorkApplyFailed,
 			Message: fmt.Sprintf("failed to build manifestwork: %v", err),
 		})
-		if updateErr := utils.PatchAddonCondition(ctx, c.addonClient, managedClusterAddonCopy, managedClusterAddon); updateErr != nil {
-			return fmt.Errorf("failed to update managedclusteraddon status: %v; the err should be %v", updateErr, err)
-		}
-		return err
+		return deployWork, hookWork, err
 	}
-	if work == nil {
-		klog.V(4).Infof("No resource needs to deploy on the managed cluster %q", key)
-		return nil
+	if deployWork != nil {
+		if owner != nil {
+			deployWork.OwnerReferences = []metav1.OwnerReference{*owner}
+		}
+		setStatusFeedbackRule(deployWork, agentAddon)
 	}
 
-	work.OwnerReferences = []metav1.OwnerReference{*owner}
+	if hookWork != nil && owner != nil {
+		hookWork.OwnerReferences = []metav1.OwnerReference{*owner}
+	}
 
-	setStatusFeedbackRule(work, agentAddon)
+	return deployWork, hookWork, nil
+}
 
+func (c *addonDeployController) applyWork(ctx context.Context, installMode string,
+	work *workapiv1.ManifestWork, addon *addonapiv1alpha1.ManagedClusterAddOn) (*workapiv1.ManifestWork, error) {
+	var appliedType string
+	switch installMode {
+	case constants.InstallModeHosted:
+		appliedType = constants.AddonHostingManifestApplied
```

Comment:
I think we should make it as an input var

Reply:
done